### PR TITLE
Add `errors` field in unsuccesful rest api responses

### LINKED ISSUE #1022: Assure all API errors are returned in json format
## Problem Definition

At the moment all error descriptions in the API are [provided in plain text](https://github.com/raiden-network/raiden/blob/master/raiden/api/rest.py#L358) as part of the RPC response. That is sometimes not what a user would expect as can be seen from https://github.com/raiden-network/raiden/issues/1020

## Solution

Make sure that all unsucesfull api responses contain an `errors` key.

## Tasklist

- [ ] Assure all errors are returned in a json object
- [ ] Update documentation of the API to reflect the change
- [ ] Add a changelog entry.

### PR DESCRIPTION
**Pull request by LefterisJP**
Fix #1022

### CODE REVIEW

File: docs/changelog.rst [hunk 1]
```diff
@@ -3,6 +3,7 @@ Changelog
 =========
 
 
+* :feature:`1022` Include ``error_msg`` in all epi responses and place result under a corresponding ``result`` key.
```
Comment (hackaugusto) [inline docs/changelog.rst]: s/epi/api/

**Comment by LefterisJP** [comment]
> Chaging all resposes to always include an error and a result key doesn't make much sense IMO. Having two different json formats is fine.

Many APIs out there have the result in a `result` key. I would like to keep it that way.

**Comment by hackaugusto** [comment]
Even though there may be other apis using the same structure, because others are doing this way is not a good motivation. This change breaks the document paradigm

**Comment by czepluch** [comment]
The changes currently look good to me. I don't have a strong opinion in regards to whether we should always include an `error` and a `result` key or use two different json formats though.

**Comment by LefterisJP** [comment]
>  This change breaks the document paradigm

Can you provide more information on what it is that is broken by this?

I believe it's beneficial because it adds uniformity to the responses of the API.

**Comment by hackaugusto** [comment]
> Can you provide more information on what it is that is broken by this?

I'm trying really hard to point out that URL are resource locators, we should use URLs as a name for a document, in our case the document may be updated and this happens as a representation of a raiden/ethereum state.

Example, `/api/1/events/channels/` should be looked at as a directory of channels, that's why it contains a list of channels and that's why its in plural, a channel is a document inside that directory, e.g. `/api/1/channels/<address>`, when that URL is requested the document that represented the channel is represented. The document changes through usage of the HTTP verbs, that is why close uses PATCH and it only sends a subset of the attributes of the document/channel.

Now, that is not perfect, but that is the closest to semantic meaning that I know off, I don't see value in have a consistent set of keys that are exclusive from one another without a semantic meaning in respect to the resource identifier.

**Comment by LefterisJP** [comment]
> I'm trying really hard to point out that URL are resource locators, we should use URLs as a name for a document

Sure and we do. Only that now the returned json document contains, in a uniform way, the expected result under `response['result']` and an optional error message under `response['error_msg']`.

I really don't see why we would make the `error_msg` appear only when there are errors and not keep a uniform API response scheme.

**Comment by LefterisJP** [comment]
I had a chat with @konradkonrad about it and he presented one more convincing argument on the approach of having two separate document types for errors and for succesful responses.

The argument is that since we use http status codes. The user should be encouraged to check for them and not for the existence of not of an `error_msg` as this way we can end up being in strange situations where we got a succesful response but somehow the error_message was also set

File: docs/changelog.rst [hunk 2]
```diff
@@ -2,6 +2,7 @@
 Changelog
 =========
 
+* :feature:`1022` Include ``error_msg`` in all api responses and place result under a corresponding ``result`` key.
```
Comment (konradkonrad) [inline docs/changelog.rst]: reword
> * :feature:`1022` Return ``errors`` field in failed API responses and place result in ``result`` field.

File: raiden/api/rest.py [hunk 3]
```diff
@@ -80,9 +80,18 @@ def normalize_events_list(old_list):
...
     response = make_response((
-        json.dumps(data),
+        json.dumps(result),
+        status_code,
+        {'mimetype': 'application/json', 'Content-Type': 'application/json'}
+    ))
+    return response
+
+
+def api_error(errors, status_code=httplib.CONFLICT):
```
Comment (konradkonrad) [inline raiden/api/rest.py]: I suggest to make `status_code` mandatory and also check for the given `status_code` to be in our (known) set of error codes:
```
ERRORS = [httplib.CONFLICT, httplib.REQUEST_TIMEOUT, ...]
def api_error(errors, status_code):
  assert status_code in ERRORS, "programming error"
  ...
```

Comment (LefterisJP) [reply raiden/api/rest.py]: okay willl do

File: docs/changelog.rst (hunk 2 above)
Comment (LefterisJP) [reply docs/changelog.rst]: Sorry this is outdated. This is no longer done. There is no `result` field.

**Comment by LefterisJP** [comment]
@konradkonrad Can you re-review. I think you misunderstood due to an outdated changelog entry. There is no result field. It's just a function argument name. In the event of a succesfull api request there is no result key, the response's `json()` is the result.

File: docs/changelog.rst (hunk 2 above)
Comment (konradkonrad) [reply docs/changelog.rst]: the changelog entry is still outdated...please adjust it according to your PR :)

Comment (LefterisJP) [reply docs/changelog.rst]: yep, will do.

**Pull request merged**